### DISCUSSION TITLE
BASED library

### DISCUSSION THREAD

**Discussion by nannerpusser** [ceccopierangiolieugenio/pyTermTk#268] [General]
Gonna be a spam bot here just to let you know this whole library and tooling is absolute money. I've been (re)learning tkinter/customtkinter for some dumpster fire personal projects lately and came across this and saw the syntax and conventions and couldn't believe it. Epic TUI stylized after the WaReZ ASCII/ANSI art scene from back in my day? Yep. Dummy like me who can understand Ttk code syntax can Plug and Play with this? Yep. I'll be sure to spam a link to whatever I end up making with it. Great work.

**Comment by ceccopierangiolieugenio** [comment]
Wow, Thank You.
Actually my first goal was to build a library that mimic the Tcl/Tk api style,
but I think that only the objects initialization may resemble it,
everything else is more close to the QT paradigms.
But still, thank you, may times I question myself due to the cold feedback I receive anytime I am trying to highlight it.
If you have some ideas or suggestion please, let me know.

**Reply by nannerpusser** [reply]
> Wow, Thank You. Actually my first goal was to build a library that mimic the Tcl/Tk api style, but I think that only the objects initialization may resemble it, everything else is more close to the QT paradigms. But still, thank you, may times I question myself due to the cold feedback I receive anytime I am trying to highlight it. If you have some ideas or suggestion please, let me know.

Still got my eye on this project 🌶️ That's a spicy demo in the README  now too!

I had a sad time trying to work with the table widget in ttkDesigner last night, but did not know it was such a recent merge. Don't want to  bother you with a full issue ticket on something so new, but if you want the tracebacks, logs, etc. I am positive I can recreate the crash 🙃 . It might just need a catcher loop on something because when I'd crash it, it would lose the terminal/screen and print the traceback to stderr over the tui that remained drawn, but couldn't exit cleanly each time and made Terminal have a seizure.

Anyway, keep up the good work, I see the updates on my main Git start page feed now, still trying to learn just enough to maybe actually be useful when I find a leet project like this.

**Reply by ceccopierangiolieugenio** [reply]
please send me more details about this crash, If I can reproduce it I will try to fix soon.
Unfortunately I feel that I need to rethink/rework the TTK Designer core routines because of the workarounds I added due to the custom nature of some widgets, so, I am expecting more edge cases where things may not works as expected.
I noticed anyway recently that the mouse press sometimes is sent to the wrong widget and the next mouse movement screw the layout many times in a non recoverable way.

thanks for your support!!!
unfortunately my dev is very slow but I try to keep it consistent, so, sooner or later I will be able to address all the issues.

**Comment by nannerpusser** [comment]
Apparently I can't actually recreate the crash when I use Remote Dev with VS Code inside WSL2... The internal term in Code should be the same regardless, but Powershell and (Windows) Terminal both updated and was how I was running ttkDesigner before. It might actually be pipx causing a problem since that's how I installed it, but I will mess with it more and make sure to grab any debug logs... But while I have your eyes momentarily, I have a couple random questions:

What (if any) alignment properties are supported by TTkLabel? I see it uses LEFT_ALIGN by definition:

`    def __init__(self, *,
                 text:TTkString="",
                 color:TTkColor=None,
                 alignment:TTkK.Alignment=TTkK.LEFT_ALIGN,
                 **kwargs) -> None:
        if issubclass(type(text), TTkString):
            self._text = text.split('\n')`

So would TTkK.RIGHT_ALIGN or CENTER work? Any Justify support on Label text or should I just make a static Text or Line widget (if possible)?

And this is really dumb, but what is the equivalent of  `anchor` or `sticky` or is there not one?

For context, I'm trying to justify or center align the text in addition to aligning the full Label widget to the right, but I think I will need to play with the grid and maybe do it that way:

```
   def __init__(self):
        super().__init__()
        self.gridLayout = ttk.TTkGridLayout(columnMinHeight=0,columnMinWidth=1)
        self.root = ttk.TTk(layout=self.gridLayout)
    
        self.create_widgets()
        items = []
    
def create_widgets(self):
        self.frame = self.gridLayout.addWidget(ttk.TTkFrame(parent=self.root, border=True), 0,1, rowspan=1, colspan=1)
        self.label = self.gridLayout.addWidget(ttk.TTkLabel(parent=self.frame, border=True, text="RESULTS", ), 0,0, rowspan=1, colspan=1)
```

**Reply by ceccopierangiolieugenio** [reply]
About JUSTIFY, the constant is there, but It is not properly implemented (it defaults to LEFT_ALIGN).
It has been in my todo-list forever, if you need it I can push its implementation in he next release.
RIGHT CENTER LEFT should work and align any string to the size of the widget, but since It was an early feature and I haven't tested it extensively, I may expect some unusual behavior.

**Reply by ceccopierangiolieugenio** [reply]
I extended your example here:

https://gist.github.com/ceccopierangiolieugenio/fb1372a71f161c82c32c6e17de2adedc

[tryItOnline](https://ceccopierangiolieugenio.github.io/pyTermTk/sandbox/sandbox.html?fileUri=https://gist.githubusercontent.com/ceccopierangiolieugenio/fb1372a71f161c82c32c6e17de2adedc/raw/b33bc0c3466332913be8a42acfc0c747b87249d3/268_Labels_Example.py)

```python
import TermTk as ttk


#   ╔══════════════════════════════════════════════════════╗
#   ║ Root widget (TTk)                                    ║
#   ║ with GridLayout this wll maximize the                ║
#   ║ frame widget to the terminal Size                    ║
#   ║ ┌──────────────────────────────────────────────────┐ ║
#   ║ │ Frame widget (TTkFrame)                          │ ║
#   ║ │ with VBoxLayout to help stacking and distribute  │ ║
#   ║ │ the labels vertically                            │ ║
#   ║ │ ┌──────────────────────────────────────────────┐ │ ║
#   ║ │ │ Label 1                                      │ │ ║
#   ║ │ │                                              │ │ ║
#   ║ │ └──────────────────────────────────────────────┘ │ ║
#   ║ │ ┌──────────────────────────────────────────────┐ │ ║
#   ║ │ │ Label 2                                      │ │ ║
#   ║ │ │                                              │ │ ║
#   ║ │ └──────────────────────────────────────────────┘ │ ║
#   ║ │  . . .                                           │ ║
#   ║ └──────────────────────────────────────────────────┘ ║
#   ╚══════════════════════════════════════════════════════╝


class Test_268():
    def __init__(self):
        super().__init__()
        self.gridLayout = ttk.TTkGridLayout()
        self.root = ttk.TTk(layout=self.gridLayout)

        self.create_widgets()
        items = []

    def create_widgets(self):
        # I am using a VBoxLayout to stack the labels vertically
        # This is a wrapper around GridLayour that helps you to avoid
        # defining the row/col,rowspan/colspan any time you add a new widget
        self.frameVBoxLayout = ttk.TTkVBoxLayout()
        self.frame = self.gridLayout.addWidget(ttk.TTkFrame(border=True, layout=self.frameVBoxLayout), 0,1)

        # Note;  in your code you were adding the labels to the root layout and not frame layout
        #        For this reason I had to define the frameVBoxLayout and add them to it
        # Note2; you don't have to specify the parent if you add a widget to the layout because
        #        it wil inherit it from the layout itself
        self.label1 = self.frameVBoxLayout.addWidget(ttk.TTkLabel(border=True, text="RESULTS\nRES1\nRes222", alignment=ttk.TTkK.Alignment.RIGHT_ALIGN))
        self.label2 = self.frameVBoxLayout.addWidget(ttk.TTkLabel(border=True, text="RESULTS\nRES1\nRes222", alignment=ttk.TTkK.Alignment.LEFT_ALIGN))
        # Or you can specify the parent and the widget will be added to its layout
        self.label3 = ttk.TTkLabel(            parent=self.frame, border=True, text="RESULTS\nRES1\nRes222", alignment=ttk.TTkK.Alignment.CENTER_ALIGN)

        # This is an exmple showing that a TTkString can be used in the label
        labelContent = ttk.TTkString('\n').join([
                ttk.TTkString("RESULTS", ttk.TTkColor.RED),
                ttk.TTkString("RESULTS Num. 1", ttk.TTkColor.BLUE + ttk.TTkColor.STRIKETROUGH),
                ttk.TTkString("RESULTS Num. 222", ttk.TTkColor.BG_GREEN),
                ttk.TTkString("RESULTS N. 3", ttk.TTkColor.BG_RED   + ttk.TTkColor.YELLOW),
            ])

        self.label4 = self.frameVBoxLayout.addWidget(ttk.TTkLabel(border=True, text=labelContent, alignment=ttk.TTkK.Alignment.CENTER_ALIGN))

    def mainLoop(self):
        self.root.mainloop()



t = Test_268()
t.mainLoop()
```
